ath79: add support for D-Link DAP-2660 A1

s-2 opened the pull request:
Specifications:
 * QCA9557, 16 MiB Flash, 128 MiB RAM, 802.11n 2T2R
 * QCA9882, 802.11ac 2T2R
 * Gigabit LAN Port (AR8035), 802.11af PoE

Installation:
 * Factory Web UI is at 192.168.0.50
   login with 'admin' and blank password, flash factory.bin
 * Recovery Web UI is at 192.168.0.50
   connect network cable, hold reset button during power-on and keep it
   pressed until uploading has started (only required when checksum is ok,
   e.g. for reverting back to oem firmware), flash factory.bin

After flashing factory.bin, additional free space can be reclaimed by
flashing sysupgrade.bin, since the factory image requires some padding
to be accepted for upgrading via OEM Web UI.

Known Issues:
 * 10BASE-T link unstable, retransmissions occur (also affects OEM fw)
 * "unknown" partition contains data, but is not declared in OEM layout

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

s-2 committed:
ath79: add support for D-Link DAP-2660 A1

Specifications:
 * QCA9557, 16 MiB Flash, 128 MiB RAM, 802.11n 2T2R
 * QCA9882, 802.11ac 2T2R
 * Gigabit LAN Port (AR8035), 802.11af PoE

Installation:
 * Factory Web UI is at 192.168.0.50
   login with 'admin' and blank password, flash factory.bin
 * Recovery Web UI is at 192.168.0.50
   connect network cable, hold reset button during power-on and keep it
   pressed until uploading has started (only required when checksum is ok,
   e.g. for reverting back to oem firmware), flash factory.bin

After flashing factory.bin, additional free space can be reclaimed by
flashing sysupgrade.bin, since the factory image requires some padding
to be accepted for upgrading via OEM Web UI.

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

s-2 commented:
Which method of cal-data extraction (.dts + `fix_wifi_mac` or `ath9k-eeprom`) would be preferred, considering the device is quite similar to DAP-2695 it might use the same case for ath9k-eeprom (by specifying `qca,no-eeprom` in the dts, otherwise the valid-looking bogus mac address from `&art 0x1002` would be used)?

There are several more devices from the the DAP-2xxx/3xxx range in the pipeline, which share a few similarities regarding the GPIO and flash layout. It's probably the cleanest and simplest solution to add support for these individual devices first, and consolidate .dtsi files afterwards?
The `"unknown"` partition contains the language pack on DAP-2695, while DAP-2660 has a separate one after the firmware partition for that, and mtd3 apparently being unused (I would even go as far as suggesting to reduce the firmware partition of DAP-2695 by 64k just to support these devices using the same .dtsi, would have to check whether this works with `mtd-fixwrgg` though). 

Images were tested for both revision A1 and A2 of this device. Would it be acceptable to use a generic variant name (e.g. Ax or just A)? I would prefer not to create two identical images for A1 and A2.

The solution of padding the factory image comes from DAP-2610 from the ipq target, and might also work for DAP-2695 A2, which would not accept the A1 image unfortunately... (but that would be for another PR)

mpratt14 commented on target/linux/ath79/dts/qca9557_dlink_dap-2660-a1.dts:
```diff
@@ -0,0 +1,151 @@
+&eth0 {
+	status = "okay";
+
+	pll-data = <0x96000000 0x94000101 0x94001313>;
```
So If speed 10 is unstable you need to try a bunch of different possible values for the cell for speed 10 which is the last one. Some possibilities replacing `0x94001313` with one of the following (other possibilities can be in other DTS files)

`0x94001616`
`0x00001313`
`0x00001616`
`0x08001313`
`0x08001616`
`0x0e001313`
`0x0e001616`
`0xa0001313`
`0xa0001616`

s-2 replied:
We've actually tried _a lot_ of values for `pll-data` settings, either dumped out of the running oem bootloader (for individual link speeds), from DD-WRT or found in other .dts files, also using various delay settings for the `&gmac-config` register (at `0x18070000`).

These efforts eventually made 1000BASE-T work for DAP-2660<sup>[0]</sup> in the first place (unlike e.g. DAP-2360<sup>[1]</sup>, which uses ar7242 that does not even have a gmac-config register, but something yet different at `0x18050004`).
The CPU PLL is 720 MHz, but the clock does not seem to be so much of an issue here (for ar7242 I had even oscilloscoped the phy pins just to verify it's actually set to 125 / 25 MHz).

However, rather than randomly brute-forcing any values that happened to work for other devices,
I would appreciate if someone could point me towards a more deterministic approach on calculating these settings.

It's curious that neither the vendor nor DD-WRT seem to get all three link speeds working smoothly,
besides I wonder why there are three competing ways to setup internal delays, and which one(s) to choose:
 * on the phy itself, using `phy-mode = "rgmii-id"` / `"rgmii-rxid"` / `"rgmii-txid"` (with `rgmii-id` just setting both rx and tx internal delays)
 * in the `pll-data` value for the particular link speed (bit [31] for inversion, [29:28] for tx delay, [27:26] for rx delay)
 * in the `gmac-config` node (only present in newer devices), there is `txd-delay` (data from gmac to phy), `txen-delay` (tx enable, i.e. clock from gmac to phy), `rxd-delay` (data from phy to gmac) and `rxdv-delay` (data valid, i.e. clock signal from phy to gmac).

Is it just me lacking the skill to see through this, or is everyone else also just fiddling with settings until it works?
(what if none of these work and we'd need to implement initialisation differently in the kernel driver etc.)

I would thus be willing to donate one of each devices to someone having more experience with how these things work in the kernel drivers (preferably in Europe to avoid customs fees and delays).


[0] https://forum.openwrt.org/t/ath79-support-d-link-dap-2660-ethernet-not-working-in-gigabit-mode/71892
[1] https://forum.openwrt.org/t/ar8021-phy-not-working-in-gigabit-mode-how-to-configure-pll-data-cells/67207/2

mpratt14 replied:
Yeah I'm one of the guys in those threads, since then I added 3 boards that need pll-data and also working on one that is not ready to add yet. I disagree that there exists a deterministic way to "find" the values you need, or at least, you need very fancy tools like oscilloscopes and a much more vast understanding of the SOC, which I don't have.

I achieved good performance on all speeds without any skills needed. Reading the registers in U-boot always had the wrong value for at least one of the speeds, and I found the rest with the guessing method. However I suppose you can read the datasheet and try to figure out which bits you need but that's a whole 32-bit word to decipher, at least half of them with a unique function.

for the 100 and 10 speeds there is always a pattern for the last bits: `0x101` for 100 and `0x1313` or `0x1616` for 10. At the very least you should try `0x94001616` if you haven't already. Good luck.

adschm commented:
> There are several more devices from the the DAP-2xxx/3xxx range in the pipeline, which share a few similarities regarding the GPIO and flash layout. It's probably the cleanest and simplest solution to add support for these individual devices first, and consolidate .dtsi files afterwards?

That typically depends on how far your work on those other devices has advanced; if you have them "ready" and know what needs to be in the DTSI files and what doesn't, add the DTSI in the first place, so we don't have to move around so much code.

On the other hand, if you have little knowledge about the other devices yet, experience is that the DTSI files will need to be changed or even redesigned later in most cases. So, in that case, I'd just add a device DTS for now.

> Images were tested for both revision A1 and A2 of this device. Would it be acceptable to use a generic variant name (e.g. Ax or just A)? I would prefer not to create two identical images for A1 and A2.

Yes, if you are absolutely sure there won't be an A3 that's different. For the DIR-842-C series, we e.g. had the C2 with USB and C1/C3 without ...

Looking at the ath79 target in particular, it seems to be common to just add the a1 variant for the name, though. This way you can point a2 to the a1, but still have the option to distinguish if a different a3 comes up.

s-2 commented:
> Looking at the ath79 target in particular, it seems to be common to just add the a1 variant for the name, though.

That was also my intention, however for DAP-2695 it would probably mean adding additional padding which is only relevant for the A2 variant, but I guess that shouldn't be an issue. Other devices (e.g. DAP-2230 and DAP-2660) need padding even for A1 revision.

Thanks for your feedback, I'll update this at the beginning of next week to move flash layout and green power led into a .dtsi.

Sharing a .dtsi with DAP-2695 would mean either reducing the firmware size, leaving 64k unallocated, or overwriting only half of the langpack partition on other devices, though I'd prefer the first option.

adschm commented:
> Sharing a .dtsi with DAP-2695 would mean either reducing the firmware size, leaving 64k unallocated, or overwriting only half of the langpack partition on other devices, though I'd prefer the first option.

Or just not having the partition layout in the DTSI ...

adschm commented:
> That was also my intention, however for DAP-2695 it would probably mean adding additional padding which is only relevant for the A2 variant, but I guess that shouldn't be an issue. Other devices (e.g. DAP-2230 and DAP-2660) need padding even for A1 revision.

What do you mean by "padding"? The factory images are different? Then we should indeed have different devices for both ...

s-2 commented:
factory images have a minimum size requirement to be flashable via oem web ui, this affects most newer devices, and also DAP-2695 A2. But the same image would still be flashable to A1 of course, the full free space is reclaimed after flashing sysupgrade.
But then again, it's probably not much of an issue either to just override the buildstep for A1 to have an image without padding, when it's not required.


Am 2. Oktober 2020 19:01, schrieb Adrian Schmutzler:

That was also my intention, however for DAP-2695 it would probably mean adding additional padding which is only relevant for the A2 variant, but I guess that shouldn't be an issue. Other devices (e.g. DAP-2230 and DAP-2660) need padding even for A1 revision.


What do you mean by "padding"? The factory images are different? Then we should indeed have different devices for both ...


—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

s-2 commented on target/linux/ath79/dts/qca95xx_dlink_dap-2xxx.dtsi:
```diff
@@ -0,0 +1,86 @@
+				label = "dlink";
+				reg = <0xf50000 0xa0000>;
+				read-only;
+			};
```
Is it acceptable to use aliases for certain partitions, to allow these to be overriden in the .dts when the flash layout for a device differs only in the ratio of these two partitions (i.e. DAP-2695) ?
Should we maybe prefix this with e.g. `mtd_` or `_partition` etc. ?

adschm replied:
Typically, I prefer to have the full partition table with the devices in the DTS files then, as everything else is typically much harder to read and thus increases the risk of having a wrong partition scheme (or updating it the wrong way later ...)

adschm replied:
You may keep part of the partition in DTSI, if only the "late" partitions differ, e.g.
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9563_tplink_tl-wpa8630p-v2.dtsi
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9563_tplink_tl-wpa8630p-v2-int.dts

s-2 commented:
So as it turns out the factory image padding will be less complicated than expected:
As stated in the wiki, the current DAP-2695 A1 image is meant to be flashed via tftp only,
since I only own A2 and it worked with additional padding, I assume this is the same for A1.

@stintel:
Since you own the A1 revision of that device, could you verify this is also flashable via A1 OEM Web UI? https://github.com/s-2/openwrt/commit/6407660bd43e8ffa2ab41a164f8455882711396e
I would then add a commit to base DAP-2695 off the same .dtsi and Makefile rule.

s-2 commented:
Moved last three partitions from .dts to .dtsi.

Turns out there are indeed more differences within the family of devices, so I would indeed refrain from creating a .dtsi cascade of hell to cover all layouts, and just have a little redundancy in the .dts instead.
```
0x000000070000-0x000000f50000: DAP-2230, 2660, 2680
0x000000070000-0x000000f60000: DAP-3320, 3662
0x000000070000-0x000000f70000: DAP-2695
0x000000070000-0x000000fe0000: DAP-2360
0x0000000a0000-0x000000fc0000: DAP-2553
```

mpratt14 commented:
@s-2 As I have said, I also have been working on boards that have an AR8035 ethernet switch. While I was reading an old pull request #1946 I had a realization and decided to take a deeper look and experimented over a couple days. You can read the latest comment of my PR for explanation #3505 

Try this configuration and let us know if it's good

```
	pll-data = <0x02000000 0x00000101 0x00001313>;

	phy-handle = <&phy0>;
	phy-mode = "rgmii-id";

	gmac-config {
		device = <&gmac>;
		rgmii-enabled = <1>;
		rxd-delay = <3>;
		rxdv-delay = <3>;
	};
```

s-2 commented:
Finally found some more time to try these settings:
> ```
> 	pll-data = <0x02000000 0x00000101 0x00001313>;
> 
> 	phy-handle = <&phy0>;
> 	phy-mode = "rgmii-id";
> 
> 	gmac-config {
> 		device = <&gmac>;
> 		rgmii-enabled = <1>;
> 		rxd-delay = <3>;
> 		rxdv-delay = <3>;
> 	};
> ```
However, neither 1000 nor 100BASE-T would work.
What should be configured for `phy-mode` in the `&mdio0` node? I had tried both `rgmii` and `rgmii-id`, but neither was successful.

Thanks a lot for pointing to the comments in other PRs, also I agree that setting the delay only in the PHY (whenever possible) should be the ideal solution.

Thus, regarding the three apparently competing ways of setting up delays, I would now conclude that
`phy-mode = "rgmii-id" / "rgmii-rxid" / "rgmii-txid"` only affects the data lines of the phy,
`pll-data` applies to the clock lines of the phy, while
`gmac-config` only affects the gmac peripheral (data and clock lines configurable separately).

So, ideally, we would only have to touch `phy-mode` and `pll-data` settings, but leave out `gmac-config` completely (if possible)?

mpratt14 commented:
@s-2 So speed 10 works?

`pll-data` is a register for GMAC0, the only thing that the at803x driver reads from is `phy-mode`

I actually got my hands on ECB1200 recently. These settings don't work for it, but rather require clock inversion, so here is one last thing to try...(with no change for speed 10 if it is working now)

```
	pll-data = <0x82000000 0x80000101 0x00001313>;

	phy-handle = <&phy0>;
	phy-mode = "rgmii-id";

	gmac-config {
		device = <&gmac>;
		rgmii-enabled = <1>;
		rxd-delay = <3>;
		rxdv-delay = <3>;
	};
```

s-2 commented:
> @s-2 So speed 10 works?

I hadn't tested after the other speeds didn't work either, but the settings with inversion look better:
1000 and 100 Mbit working, but for 10 it's still not perfect, the transfer speed in iperf3 varies between 5 and 10 Mbit/s). But it seems quite stable, so I'll probably push these settings and rebase.

However I still don't even trust my test setup... USB 3.0 Gigabit adapter (Realtek) will not care about the settings in ubuntu network manager, but negotiate 1000BASE-T, internal network card seems to use half-duplex only whenever the setting is not 'automatic' etc... I'll just purchase an old 10Mbit hub, and maybe also test on Windows again.

Thanks for your help! I think the settings at least look better by using `gmac-config` delays instead of `pll-data`.

mpratt14 commented:
@s-2 So if you didn't test speed 10 probably its better to put inverted clock on that too

`	pll-data = <0x82000000 0x80000101 0x80001313>;`

FYI I also have a Realtek based USB to Ethernet adapter and I have found changing speed to be very reliable on Windows using
`ncpa.cpl` --> Properties --> Configure --> Advanced

this is the IDs for mine
VID=0BDA
PID=8153

stintel commented:
> Since you own the A1 revision of that device, could you verify this is also flashable via A1 OEM Web UI? [s-2@6407660](https://github.com/s-2/openwrt/commit/6407660bd43e8ffa2ab41a164f8455882711396e)
> I would then add a commit to base DAP-2695 off the same .dtsi and Makefile rule.

Is this question still relevant?

s-2 commented:
> Is this question still relevant?

Actually, yes - I only have revision A2 of this device, but would expect the image formats to be identical (D-Link only provide a single Ax image). However I would prefer not to commit changes to a device without verifying on real hardware.

s-2 commented:
rebased to master, changed gmac pll settings (10BASE-T still not perfect but workign okay)

mpratt14 commented:
I suppose for speed 10 you can try

`0x80001616`

but I think this is more likely to break it...probably this is the best we can do

s-2 commented:
> `0x80001616`
Tried that already, but no throughput at all, so `0x80001313` is probably the best choice for now.

mpratt14 commented:
Yeah at this point it looks like we have the best register values, and maybe this is caused by a small issue with the PCB design

pinging with large packets is clean output or is there packet loss?

s-2 commented:
Just realized the order of `model` and `compatible` properties in the .dts is quite random; changed to have `compatible` first to be consistent with DAP-2695, in preparation of later merging with .dtsi.

adschm commented on target/linux/ath79/dts/qca95xx_dlink_dap-2xxx.dtsi:
```diff
@@ -0,0 +1,68 @@
+&spi {
+	status = "okay";
+
+	num-cs = <1>;
```
num-cs can probably be dropped.

s-2 replied:
I'll just leave this here as I have no account on patchwork:
Some ath79 devices also still have `cs-gpios = <0>, <0>;` along with num-cs, which should be removed as well (no idea why someone would declare two chipselects on the same gpio).

adschm replied:
That's a different subject, but you could patch that out as well in a separate patch if you like.

s-2 commented:
- removed `num-cs` from `&spi` node
- removed resolved 10BASE-T link speed issue from commit message
- also removed `phy-mask` from `&mdio0` node (seems to be another property being used inconsistently / copied randomly from other devices; apparently used to suppress auto-probing for mdio phy addresses (0-31) matching the corresponding bit positions; but not relevant for this device)
- rebased to master

s-2 commented:
* moved reset button to individual .dts file (in foresight of further devices from the series to be added)
* added `default-state` for `led_power` to avoid ~2 seconds gap in boot led pattern
* rebased to master

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

s-2 commented:
Awesome, thanks!

How shall we proceed with additional devices from the series?
Currently I have a few prepared as separate local branches based on this one (waiting for review to be completed first, to avoid having to fix identical issues separately for multiple devices).

I guess it is probably easiest to merge them into a single new branch based on the next master (with DAP-2660 included), to minimize merging hassle (mostly from `base-files` entries).

adschm commented:
> Awesome, thanks!
> 
> How shall we proceed with additional devices from the series?
> Currently I have a few prepared as separate local branches based on this one (waiting for review to be completed first, to avoid having to fix identical issues separately for multiple devices).
> 
> I guess it is probably easiest to merge them into a single new branch based on the next master (with DAP-2660 included), to minimize merging hassle (mostly from `base-files` entries).

One PR for many devices might be okay if differences between devices are trivial, but please have one commit per device.

s-2 commented:
Opened #3707 adding 4 additional devices, that's all I have for the moment (aside from ancient DAP-2553 and 2360, which seem to require further investigation regarding GMAC settings on ar9xxxx SoC's).
Newer devices (DAP-2620, 2662, 2682, ...) use a different factory format or even new SoCs (QCN5502), which require further investigation.
